File: src/type/permissions.ts
/**
 * RBAC Permission Types
 * 
 * Type definitions for role-based access control from the backend.
 */

/** Per-model permission configuration */
export interface ModelPermissions {
  /** Whether user can view records of this model */
  view: boolean;
  /** Fields user can view, or ["*"] for all */
  view_fields: string[];
  /** Whether user can edit records of this model */
  edit: boolean;
  /** Fields user can edit, or ["*"] for all */
  edit_fields: string[];
  /** Whether user can create new records */
  create: boolean;
  /** Whether user can delete records */
  delete: boolean;
}

/** User's complete permissions response from /wcapi/permissions/ */
export interface UserPermissions {
  /** Django user ID */
  user_id: number;
  /** User's assigned roles */
  roles: string[];
  /** Org IDs user is associated with by type */
  org_ids: {
    customer: number[];
    vendor: number[];
    manufacturer: number[];
    employee: number[];
  };
  /** Contact record ID if linked */
  contact_id: number | null;
  /** Whether user is a superuser */
  is_superuser: boolean;
  /** Per-model permissions */
  models: Record<string, ModelPermissions>;
}

/** Role identifiers */
export type RoleName =
  | "user_customer"
  | "user_vendor"
  | "user_manufacturer"
  | "user_rep"
  | "user_sales"
  | "user_production"
  | "user_accounting"
  | "user_warehouse"
  | "admin"
  | "superuser";

/** Common model names for type safety */
export type ModelName =
  | "order"
  | "invoice"
  | "proposal"
  | "purchase"
  | "workorder"
  | "customer"
  | "vendor"
  | "manufacturer"
  | "contact"
  | "action"
  | "document"
  | "item"
  | "category";

/** Check if a field is in allowed list (handles "*" wildcard) */
export function isFieldAllowed(
  field: string,
  allowedFields: string[]
): boolean {
  if (!allowedFields || allowedFields.length === 0) {
    return false;
  }
  
  if (allowedFields.includes("*")) {
    return true;
  }
  
  // Exact match
  if (allowedFields.includes(field)) {
    return true;
  }
  
  // Parent path allows children (e.g., "totals" allows "totals.total")
  const parts = field.split(".");
  for (let i = 0; i < parts.length; i++) {
    const parent = parts.slice(0, i + 1).join(".");
    if (allowedFields.includes(parent)) {
      return true;
    }
  }
  
  return false;
}

/** Check if user has any of the specified roles */
export function hasRole(
  permissions: UserPermissions | null,
  roles: RoleName | RoleName[]
): boolean {
  if (!permissions) return false;
  if (permissions.is_superuser) return true;
  
  const roleArray = Array.isArray(roles) ? roles : [roles];
  return roleArray.some((role) => permissions.roles.includes(role));
}

/** Check if user has access to a model for a specific operation */
export function hasModelAccess(
  permissions: UserPermissions | null,
  model: string,
  operation: "view" | "edit" | "create" | "delete"
): boolean {
  if (!permissions) return false;
  if (permissions.is_superuser) return true;
  
  const modelPerms = permissions.models[model];
  if (!modelPerms) return false;
  
  return modelPerms[operation] as boolean;
}

/** Check if user can view a specific field on a model */
export function canViewField(
  permissions: UserPermissions | null,
  model: string,
  field: string
): boolean {
  if (!permissions) return false;
  if (permissions.is_superuser) return true;
  
  const modelPerms = permissions.models[model];
  if (!modelPerms || !modelPerms.view) return false;
  
  return isFieldAllowed(field, modelPerms.view_fields);
}

/** Check if user can edit a specific field on a model */
export function canEditField(
  permissions: UserPermissions | null,
  model: string,
  field: string
): boolean {
  if (!permissions) return false;
  if (permissions.is_superuser) return true;
  
  const modelPerms = permissions.models[model];
  if (!modelPerms || !modelPerms.edit) return false;
  
  return isFieldAllowed(field, modelPerms.edit_fields);
}

File: backend/apps/core/views/permissions.py
"""
RBAC Permissions API Views.

Endpoints for retrieving user permissions and role information.
"""
from __future__ import annotations

from typing import Any, Dict

from rest_framework import status
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response
from rest_framework.views import APIView

from apps.core.services.role_filter import (
    build_user_context,
    get_user_filter_config,
    can_create,
    can_delete,
    get_allowed_fields,
)


class UserPermissionsView(APIView):
    """
    Get current user's effective permissions.
    
    Returns role information and per-model access controls.
    
    GET /wcapi/permissions/
    Response:
    {
        "roles": ["user_sales", "user_production"],
        "org_ids": {
            "customer": [],
            "vendor": [],
            "manufacturer": [],
            "employee": [1]
        },
        "contact_id": 42,
        "is_superuser": false,
        "models": {
            "order": {
                "view": true,
                "view_fields": ["*"] or ["id", "ida", ...],
                "edit": true,
                "edit_fields": ["*"] or ["status", ...],
                "create": true,
                "delete": false
            },
            ...
        }
    }
    """
    
    permission_classes = [IsAuthenticated]
    
    # Models to include in permissions response
    INCLUDED_MODELS = [
        # Transactions
        "order", "invoice", "proposal", "purchase", "workorder",
        # Orgs
        "customer", "vendor", "manufacturer",
        # Core
        "contact", "action", "document",
        # Products
        "item", "category",
    ]
    
    def get(self, request, *args, **kwargs):
        user = request.user
        
        # Build user context
        context = build_user_context(user)
        
        # Build per-model permissions
        models_perms: Dict[str, Dict[str, Any]] = {}
        
        for model_name in self.INCLUDED_MODELS:
            config = get_user_filter_config(user, model_name)
            
            if config:
                view_fields = config.get("view_fields", [])
                edit_fields = config.get("edit_fields", [])
                
                models_perms[model_name] = {
                    "view": bool(view_fields),
                    "view_fields": view_fields,
                    "edit": bool(edit_fields),
                    "edit_fields": edit_fields,
                    "create": config.get("allow_create", False),
                    "delete": config.get("allow_delete", False),
                }
            else:
                # No config = no access
                models_perms[model_name] = {
                    "view": False,
                    "view_fields": [],
                    "edit": False,
                    "edit_fields": [],
                    "create": False,
                    "delete": False,
                }
        
        response_data = {
            "user_id": context.get("user_id"),
            "roles": context.get("roles", []),
            "org_ids": context.get("org_ids", {}),
            "contact_id": context.get("contact_id"),
            "is_superuser": context.get("is_superuser", False),
            "models": models_perms,
        }
        
        return Response(response_data, status=status.HTTP_200_OK)


class ModelPermissionsView(APIView):
    """
    Get permissions for a specific model.
    
    GET /wcapi/permissions/<model_name>/
    Response:
    {
        "model": "order",
        "view": true,
        "view_fields": ["*"],
        "edit": true,
        "edit_fields": ["status", "notes"],
        "create": true,
        "delete": false,
        "query_filters": {...}
    }
    """
    
    permission_classes = [IsAuthenticated]
    
    def get(self, request, model_name=None, *args, **kwargs):
        if not model_name:
            return Response(
                {"detail": "model_name required"},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        user = request.user
        config = get_user_filter_config(user, model_name)
        
        if not config:
            return Response({
                "model": model_name,
                "view": False,
                "view_fields": [],
                "edit": False,
                "edit_fields": [],
                "create": False,
                "delete": False,
                "query_filters": {},
            }, status=status.HTTP_200_OK)
        
        return Response({
            "model": model_name,
            "view": bool(config.get("view_fields")),
            "view_fields": config.get("view_fields", []),
            "edit": bool(config.get("edit_fields")),
            "edit_fields": config.get("edit_fields", []),
            "create": config.get("allow_create", False),
            "delete": config.get("allow_delete", False),
            "query_filters": config.get("query_filters", {}),
        }, status=status.HTTP_200_OK)
